api: add sata controller

tenthirtyam opened the pull request:
## Description

Adds support for creating SATA controllers.

Closes: #1219

## Type of change

Please mark options that are relevant:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] This change requires a documentation update
- [ ] Build related change

## How Has This Been Tested?

- [x] Local test: `TestCreateSATAController`

```sh
Running tool: /usr/local/bin/go test -timeout 30s -run ^TestCreateSATAController$ github.com/vmware/govmomi/object

ok  	github.com/vmware/govmomi/object	(cached)
```

## Checklist:

- [x] My code follows the `CONTRIBUTION` [guidelines] of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged

[guidelines]: https://github.com/vmware/govmomi/blob/main/CONTRIBUTING.md

dougm reviewed (approved):
lgtm, thanks @tenthirtyam !

akutz reviewed (dismissed):
Did you account for [`VirtualAHCIController`](https://vdc-download.vmware.com/vmwb-repository/dcr-public/184bb3ba-6fa8-4574-a767-d0c96e2a38f4/ba9422ef-405c-47dd-8553-e11b619185b2/SDK/vsphere-ws/docs/ReferenceGuide/vim.vm.device.VirtualAHCIController.html)?

dougm commented on object/virtual_device_list.go:
```diff
@@ -361,6 +373,63 @@ func (l VirtualDeviceList) newNVMEBusNumber() int32 {
+		return nil, fmt.Errorf("%s is not an SATA controller", name)
+	}
+
+	c := l.PickController((*types.VirtualSATAController)(nil))
```
There shouldn't be a panic here. `l` is the VirtualDeviceList and the param (nil pointer here) passed to `PickController` is just used for its type, used to find an existing instance of that type.

akutz replied:
Maybe not a panic, but `c` may be `nil` when it does not need to be based on sending in `*types.VirtualSATAController`. This does not allow for AHCI here.

tenthirtyam replied:
Should beresolved in https://github.com/vmware/govmomi/commit/57aea68935dca95a7bb8f73c3ffb4b9905753861.

dougm commented on object/virtual_device_list.go:
```diff
@@ -389,6 +460,8 @@ func (l VirtualDeviceList) PickController(kind types.BaseVirtualController) type
 			return num < 15
 		case *types.VirtualIDEController:
 			return num < 2
+		case *types.VirtualSATAController:
```
:+1: , I didn't realize SATA was extended by AHCI. Worthwhile to include the AHCI case(s) in this PR too.

tenthirtyam replied:
Should be resolved in https://github.com/vmware/govmomi/commit/57aea68935dca95a7bb8f73c3ffb4b9905753861.

tenthirtyam commented:
I think that I've addressed your concerns, but let me know otherwise.

I still want to do a round of additional tests before merge.

dougm commented on object/virtual_device_list.go:
```diff
@@ -911,6 +998,8 @@ func (l VirtualDeviceList) Type(device types.BaseVirtualDevice) string {
 		return "lsilogic-sas"
 	case *types.VirtualNVMEController:
 		return "nvme"
+	case *types.VirtualSATAController:
```
No need for this, see `deviceNameRegexp`. Realizing we don't need the existing NVME case either.

tenthirtyam replied:
Want me to remove nvme or add that in a separate PR?

tenthirtyam replied:
Resolved in 2e77836.

dougm commented on object/virtual_device_list_test.go:
```diff
@@ -762,6 +762,20 @@ func TestCreateSCSIController(t *testing.T) {
+	for _, l := range []VirtualDeviceList{SATAControllerTypes(), devices} {
+		_, err := l.CreateSATAController()
+		if err != nil {
+			t.Error("should fail")
```
Looks like some copy-n-pasta from other tests, but can be removed here.

tenthirtyam replied:
Resolved in 2e77836.

dougm commented on object/virtual_device_list.go:
```diff
@@ -361,6 +373,77 @@ func (l VirtualDeviceList) newNVMEBusNumber() int32 {
+	case *types.VirtualSATAController:
+		return c, nil
+	case *types.VirtualAHCIController:
+		return &c.VirtualSATAController, nil
```
This should return `c`, not a pointer to the embedded field, right?

tenthirtyam replied:
Resolved in 2e77836.

dougm commented on object/virtual_device_list.go:
```diff
@@ -361,6 +373,77 @@ func (l VirtualDeviceList) newNVMEBusNumber() int32 {
+		case *types.VirtualSATAController:
+			return c, nil
+		case *types.VirtualAHCIController:
+			return &c.VirtualSATAController, nil
```
This should return `c`, not a pointer to the embedded field, right?

tenthirtyam replied:
Resolved in 2e77836.

dougm commented on object/virtual_device_list.go:
```diff
@@ -57,6 +57,18 @@ func SCSIControllerTypes() VirtualDeviceList {
+func SATAControllerTypes() VirtualDeviceList {
+	// Return a mutable list of SATA controller types, initialized with defaults.
+	return VirtualDeviceList([]types.BaseVirtualDevice{
+		&types.VirtualSATAController{},
```
If this function were used, should also have (or only have?) `VirtualAHCIController` in this list?

tenthirtyam replied:
Resolved in 2e77836.

dougm commented on govc/vm/create.go:
```diff
@@ -519,6 +519,14 @@ func (cmd *create) addStorage(devices object.VirtualDeviceList) (object.VirtualD
 			devices = append(devices, nvme)
 			cmd.controller = devices.Name(nvme)
+		} else if cmd.controller == "sata" {
+			sata, err := devices.CreateSATAController()
```
Just trying in the H5 UI for the first time..
`Add New Device -> Controllers -> SATA Controller`
Always creates a `VirtualAHCIController`, I don't see an option to create the base `VirtualSATAController` type.
Seems we should be doing the same, any insights on this?

```console
% govc device.info -vm my-vm ahci-*
Name:       ahci-15000
  Type:     VirtualAHCIController
  Label:    SATA controller 0
  Summary:  AHCI
  Key:      15000
  Devices:  
```

dougm replied:
Tried creating a vm with your branch:
```console
% govc vm.create -disk.controller sata -net "VM Network" -host $host -ds $ds sata-vm
govc: A specified parameter was not correct:  (Copying Virtual Machine configuration)
```
Works with this change:
```diff
modified   object/virtual_device_list.go
@@ -412,7 +413,7 @@ func (l VirtualDeviceList) FindSATAController(name string) (types.BaseVirtualCon
 
 // CreateSATAController creates a new SATA controller.
 func (l VirtualDeviceList) CreateSATAController() (types.BaseVirtualDevice, error) {
-	sata := &types.VirtualSATAController{}
+	sata := &types.VirtualAHCIController{}
 	sata.BusNumber = l.newSATABusNumber()
 	sata.Key = l.NewKey()
```

tenthirtyam replied:
Resolved in 2e77836.

dougm commented on object/virtual_device_list.go:
```diff
@@ -57,6 +57,18 @@ func SCSIControllerTypes() VirtualDeviceList {
 }
 
+// SATAControllerTypes are used for adding a new SATA controller to a VM.
+func SATAControllerTypes() VirtualDeviceList {
```
Don't see this function used anywhere, seems we don't need it?

tenthirtyam replied:
It's used in the `TestCreateSATAController` in the tests.

tenthirtyam replied:
Resolved in 2e77836.

dougm commented on object/virtual_device_list.go:
```diff
@@ -361,6 +373,77 @@ func (l VirtualDeviceList) newNVMEBusNumber() int32 {
+
+// CreateSATAController creates a new SATA controller.
+func (l VirtualDeviceList) CreateSATAController() (types.BaseVirtualDevice, error) {
+	sata := &types.VirtualSATAController{}
```
See previous comment regarding UI, seems this should be `VirtualAHCIController` ?

tenthirtyam replied:
Resolved in 2e77836.

dougm reviewed (commented):
A few things I didn't notice on the 1st pass, esp. after having tried from the UI

tenthirtyam commented:
> A few things I didn't notice on the 1st pass, esp. after having tried from the UI

Thanks, Doug! I've addressed your comments - appreciate your help!

dougm reviewed (approved):
lgtm, thanks @tenthirtyam
